Indexing/LiteralSubstitutionTree.hpp

```diff
@@ -36,134 +47,27 @@ class LiteralSubstitutionTree
-      if(_retrieveSubstitutions) {
-        return instantiation ? res.substitution->matchSorts(_queryEqSort, resSort) 
-                             : res.substitution->matchSorts(resSort, _queryEqSort); 
-      } else {
-        static RobSubstitution subst;
-        subst.reset();
-        return instantiation ? subst.match(_queryEqSort, 0, resSort, 1):
-                               subst.match(resSort, 0, _queryEqSort, 1);           
-      }
-    }
-
-    //dummy. UnificationFilter needs a leave function to undo the sort unification.
-    //MatchingFilter doesn't require this, since the sort unifier is added onto
-    //the final term unifier and undone by the next call to backTrack() in FastGen 
-    //or FastInst iterator.
-    void leave(const SLQueryResult& res){  }
-  private:
-    TermList _queryEqSort;
-    bool _isTwoVarEq;
-    bool _retrieveSubstitutions;
-  };
-
-  template <bool polymorphic>
-  struct UnificationFilter
-  {
-    UnificationFilter(Literal* queryLit, bool retrieveSubstitutions)
-    : _queryEqSort(SortHelper::getEqualityArgumentSort(queryLit)), 
-      _retrieveSubs(retrieveSubstitutions) {}
-
-    bool enter(const SLQueryResult& res)
-    {
-      CALL("LiteralSubstitutionTree::UnificationFilter::enter()");
-      ASS(res.literal->isEquality());
-      
-      //the polymorphism check isn't strictly necessary. However, if it wasn't
-      //included, on monomorphic problems we would be using unification to check
-      //whether two constant are identical
-
-      TermList resSort = SortHelper::getEqualityArgumentSort(res.literal);
-      if(!polymorphic){
-        return _queryEqSort == resSort;
-      } else if(_retrieveSubs){
-        RobSubstitution* subst = res.substitution->tryGetRobSubstitution();
-        ASS(subst);
-        subst->bdRecord(_bdataEq);
-        bool success = subst->unify(_queryEqSort, 0, resSort, 1);
-        subst->bdDone();
-        if(!success){
-          _bdataEq.backtrack();
-        }
-        return success;
-      } else {
-        static RobSubstitution subst;
-        subst.reset();
-        return subst.unify(_queryEqSort, 0, resSort, 1);
-      }
-    }
-
-    void leave(const SLQueryResult& res){
-      CALL("LiteralSubstitutionTree::UnificationFilter::leave()");
-      if(_retrieveSubs && polymorphic){
-        _bdataEq.backtrack();
-        ASS(_bdataEq.isEmpty());
-      }
-    }
-  private:
-    TermList _queryEqSort;
-    bool _retrieveSubs;
-    BacktrackData _bdataEq;
-  };
-
-  template<class Iterator, class Filter>
-  SLQueryResultIterator getResultIterator(Literal* lit,
-	  bool complementary, bool retrieveSubstitutions, bool useConstraints);
-
-  unsigned getRootNodeIndex(Literal* t, bool complementary=false);
-  bool _polymorphic;
+  template<class Iterator>
+  SLQueryResultIterator getResultIterator(Literal* lit, bool complementary, bool retrieveSubstitutions, bool useConstraints);
+
+  Stack<unique_ptr<SubstitutionTree>> _trees;
```

Comment:
Can this just be a `Stack<SubstitutionTree>`?

Reply:
I know `getTree` returns a reference that would be invalidated on resize, but it seems like we shouldn't resize while the reference is live.